DMCompiler/DM/Visitors/DMObjectBuilder.cs

```diff
@@ -1,105 +1,142 @@
 using OpenDreamShared.Compiler;
 using DMCompiler.Compiler.DM;
 using OpenDreamShared.Dream;
-using System;
 using OpenDreamShared.Dream.Procs;
 using System.Collections.Generic;
-using Robust.Shared.Utility;
 
 namespace DMCompiler.DM.Visitors {
-    static class DMObjectBuilder {
-        /// <summary>
-        /// In DM, the definition of a base class may occur way after the definition of (perhaps numerous) derived classes. <br/>
-        /// At the time that we first evaluate the derived class, we do not know some important information, like the implicit type of certain things.<br/>
-        /// This event allows for delaying variable override evaluation until after we know what it is. </summary>
-        /// <remarks>
-        /// Further, it just so happens to also act as a container that can enumerate all objects which were expecting a definition but never got one.
-        /// </remarks>
-        public static event EventHandler<DMVariable>? VarDefined; // Fires if we define a new variable.
-
-        /// <summary>
-        /// A running collection of types that we need to find a definition for. <br/>
-        /// Comes up most especially when a var/some/type/x is defined before we know what /some/type even is. <br/>
-        /// The Location value can just be the first instance of this definition being promised, for errors' sake.
-        /// </summary>
-        public static Dictionary<DreamPath, Location> AwaitedObjectDefinitions = new();
-
-        /// <summary>
-        /// Similiar to <see cref="AwaitedObjectDefinitions"/> except for procs that are overridden before they are defined.
-        /// </summary>
-        public static Dictionary<string, DMObject> AwaitedProcDefinitions = new();
+    internal static class DMObjectBuilder {
+        private static readonly List<(DMObject, DMASTObjectVarDefinition)> VarDefinitions = new();
+        private static readonly List<(DMObject, DMASTObjectVarOverride)> VarOverrides = new();
+        private static readonly List<(DMObject?, DMASTProcDefinition)> ProcDefinitions = new();
+        private static readonly List<(DMObject DMObject, DMProc Proc, int Id, DMASTProcStatementVarDeclaration VarDecl)> StaticProcVars = new();
```

Comment:
For the little perf bonus, you could check how large these lists get when compiling a target and pre-allocate when DMObjectBuilder is instantiated.

Reply:
A profile shows ~1ms spent adding to these lists. I don't think it's a concern.